### PR TITLE
various sd-Ipv4ll/sd-ipv4acd fixes

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by teg** [comment]
Looks all good to me!

File: src/libsystemd-network/sd-ipv4acd.c [hunk 1]
```diff
@@ -71,14 +71,19 @@ struct sd_ipv4acd {
         IPv4ACDState state;
         int ifindex;
         int fd;
-        int iteration;
-        int conflict;
+
+        unsigned n_iteration;
+        unsigned n_conflict;
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4acd.c:72]: I'm not sure if this rename is beneficial. We use "n_" prefix when we have a variable "somethings", and need a counter to interpret it "n_somethings". "Iteration" carries an inherent meaning of counting by itself.

File: src/libsystemd-network/sd-ipv4ll.c [hunk 2]
```diff
@@ -202,7 +200,7 @@ int sd_ipv4ll_set_address_seed(sd_ipv4ll *ll, unsigned seed) {
         if (!random_data_state)
                 return -ENOMEM;
 
-        r = initstate_r(seed, random_data_state, 128, random_data);
+        r = initstate_r((unsigned) seed, random_data_state, 128, random_data);
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4ll.c]: Hm, doesn't this just push the problem around? If initstate_r takes an unsigned, which is shorter than uint64_t (at least on some arches), a certain value of seed specified for sd_ipv5ll_set_address_seed will yield different behaviours. So maybe it's better to bite the bullet and specify the size of the seed as the _minimum_ of available sizes, i.e. uint32_t, to have consistent interpretation of the argument on all architectures. After all 2^32 should be enough seeds.

Comment (keszybz) [inline src/libsystemd-network/sd-ipv4ll.c]: Oh, you change the hashing function in the next commit. That's fine that.

File: src/libsystemd-network/sd-ipv4ll.c [hunk 3]
```diff
@@ -254,48 +228,64 @@ int sd_ipv4ll_set_address(sd_ipv4ll *ll, const struct in_addr *address) {
...
-        if (r < 0)
-                return r;
+                /* Increase the generation counter by one */
+                ll->seed.generation = htole64(le64toh(ll->seed.generation) + 1);
 
-        return 0;
+                addr = htobe32((h & UINT32_C(0x0000FFFF)) | IPV4LL_NETWORK);
+        } while (addr == ll->address ||
+                (be32toh(addr) & 0x0000FF00) == 0x0000 ||
+                (be32toh(addr) & 0x0000FF00) == 0xFF00);
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4ll.c]: `IN_SET(be32toh(addr) & 0x0000FF00, 0x0000, 0xFF00)`?

File: src/libsystemd-network/sd-ipv4ll.c [hunk 4]
```diff
@@ -248,6 +252,9 @@ static int ipv4ll_pick_address(sd_ipv4ll *ll) {
                 (be32toh(addr) & 0x0000FF00) == 0x0000 ||
                 (be32toh(addr) & 0x0000FF00) == 0xFF00);
 
+        (void) in_addr_to_string(AF_INET, &(union in_addr_union) { .in.s_addr = addr }, &address);
+        log_ipv4ll(ll, "Picked new IP address %s.", address);
```
Comment (keszybz) [inline src/libsystemd-network/sd-ipv4ll.c]: `strnull(address)`

File: src/libsystemd-network/sd-ipv4acd.c (hunk 1 above)
Comment (poettering) [inline src/libsystemd-network/sd-ipv4acd.c:72]: well, iteration, does, but conflict does not...

I usually used the "n_" prefix whenever you count something, i.e. you regularly see a line like `n_foo++;` in place

File: src/libsystemd-network/sd-ipv4ll.c (hunk 3 above)
Comment (poettering) [inline src/libsystemd-network/sd-ipv4ll.c]: will add a fix for this on top.

File: src/libsystemd-network/sd-ipv4ll.c (hunk 4 above)
Comment (poettering) [inline src/libsystemd-network/sd-ipv4ll.c]: indeed, will fix.

**Comment by poettering** [comment]
I force pushed a new version now, with a number of additional patches with fixes. Also made two of the three changes you suggested. Please have a look and merge!

**Pull request merged**